pyi.py

```diff
@@ -708,6 +682,96 @@ def _analyse_typing_Literal(node: ast.Subscript) -> TypingLiteralAnalysis:
     )
 
 
+_KNOWN_ENUM_BASES = frozenset(
+    {"Enum", "Flag", "IntEnum", "IntFlag", "StrEnum", "ReprEnum"}
+)
+
+
+_COMMON_METACLASSES = {
+    "type": "builtins",
+    "ABCMeta": "abc",
+    "EnumMeta": "enum",
+    "EnumType": "enum",
+}
+
+
+@dataclass(frozen=True)
+class EnclosingClassContext:
+    node: ast.ClassDef
+    cls_name: str
+    bases_map: defaultdict[str, set[str | None]]
+
+    def contains_in_bases(self, obj: str, *, from_: AbstractSet[str]) -> bool:
+        if obj not in self.bases_map:
+            return False
+        if None in self.bases_map[obj]:
+            return True
+        return bool(self.bases_map[obj] & from_)
+
+    @cached_property
+    def is_protocol_class(self) -> bool:
+        return self.contains_in_bases("Protocol", from_=_TYPING_MODULES)
+
+    @cached_property
+    def is_typeddict_class(self) -> bool:
+        return self.contains_in_bases(
+            "TypedDict", from_=(_TYPING_MODULES | {"mypy_extensions"})
+        )
+
+    @cached_property
+    def is_enum_class(self) -> bool:
+        return any(
+            self.contains_in_bases(enum_cls, from_={"enum"})
+            for enum_cls in _KNOWN_ENUM_BASES
+        )
+
+    @cached_property
+    def is_metaclass(self) -> bool:
+        return any(
+            self.contains_in_bases(metacls, from_={module})
+            for metacls, module in _COMMON_METACLASSES.items()
+        )
+
+    @cached_property
+    def is_decorated_with_final(self) -> bool:
+        return any(_is_final(decorator) for decorator in self.node.decorator_list)
+
+
+class ClassBase(NamedTuple):
+    module: str | None
+    obj: str
+
+
+def _analyze_classdef(node: ast.ClassDef) -> EnclosingClassContext:
+    bases_map: defaultdict[str, set[str | None]] = defaultdict(set)
+
+    def _unravel(node: ast.expr) -> str | None:
+        if isinstance(node, ast.Name):
+            return node.id
+        if isinstance(node, ast.Attribute):
+            return f"{_unravel(node.value)}.{node.attr}"
+        return None
+
+    def _analyze_base_node(
+        base_node: ast.expr, top_level: bool = True
+    ) -> ClassBase | None:
+        if isinstance(base_node, ast.Name):
+            return ClassBase(None, base_node.id)
+        if isinstance(base_node, ast.Attribute):
+            return ClassBase(_unravel(base_node.value), base_node.attr)
+        if isinstance(base_node, ast.Subscript) and top_level:
+            return _analyze_base_node(base_node.value, top_level=False)
+        return None
```

Comment:
Same question here. (A subscript can't be a Subscript's value, I think.)

Reply:
Well, it is *possible*...

(I'll admit I had to try quite hard here!)

```pycon
Python 3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> class Strange:
...     def __class_getitem__(cls, item):
...         return Strange
...
>>> class Foo(Strange[Strange][Strange]): ...
...
>>> import ast
>>> tree = ast.parse("class Foo(Strange[Strange][Strange]): ...")
>>> tree.body[0].bases[0]
<ast.Subscript object at 0x000001E44FC43FD0>
>>> _.value
<ast.Subscript object at 0x000001E44FC43F90>
```